### ISSUE TITLE
GPU Detection in runtime Selfcheck

### ISSUE THREAD

**Issue by golmek** [golemfactory/ya-runtime-vm#157]
Create self-check image (<300mb) for GPU

**Prerequisite**
https://github.com/golemfactory/ya-runtime-vm/issues/161

**Acceptance Criteria:**

- [x] Create self-check image for GPU path, which will convey information about GPU support to offer via self-check output.
   -  ~or potentially it can be the same image with different parameters -> TO BE DECIDED during development.~
- [x] Working flow for Offer to contain information about GPU
- [x] Based on the implementation, fill out the table (GPU Offer extension wishlist) with information if this data is possible to be put in offer. Table is just a wishlist of parameters which would be great to have in a offer, but it is not requirement.
- [x] blocked by an issue with the runtime, see https://github.com/golemfactory/ya-runtime-vm/issues/157#issuecomment-1613896265, https://github.com/golemfactory/ya-runtime-vm/issues/168.
- [ ] fill out `capabilities` depending on self-check output.
- [ ] move self-test output from `golem.runtime.!exp.inf` to `golem.!exp.gap-35.v1.inf`

**Technical Details:**

based on https://github.com/golemfactory/ya-runtime-vm/issues/162

- [ ] size <300mb
  - current size (based on `maugnorbert/docker_golem_cuda_base:latest`) is 577MiB. 
  - won't happen.
- [x] ~check if `/dev/nvidia*` exists.~
  - this failure mode is detected with `nvidia-smi`. 
- [x] find information about GPU card model & memory (nvidia-smi ?)
- [ ] ~(optional) run some simple cuda task.~
- [x] return GPU parameters as a output of healthcheck
- [x] clarify exact list of parameters with arek

**GPU Offer extension wishlist:**

Priority | Information | Type | Comment
-- | -- | -- | --
1 | GPU model | string | 
1 | CUDA enabled | bool | this is related to the image running and not the VM runtime
1 | RAM (dedicated video memory) | int |  
2 | RAM (total video memory) | int
2 | CUDA version compatibility | int | this is not known by the GPU because there is a minimum CUDA version and also it can be deprecated in a future CUDA version. It can also be different for CUDA and a given library, for example tensorflow had a different GPU compatibility requirement for building from source and using prebuild binaries. Get the info from [Nvidia](https://developer.nvidia.com/cuda-gpus) and/or a general guideline from [StackOverflow](https://stackoverflow.com/questions/28932864/which-compute-capability-is-supported-by-which-cuda-versions)
2 | CUDA cores | int  | This can be found from the general specification of each GPU. Otherwise it is platform specific that we would not like to work with. We should aim to focus on platform independent metrics that are easy to access in a health check.
2 | Base clock | int
2 | Boost clock | int
2 | Memory clock | int
3 | Memory type | string / int
3 | Memory bus | string / int
3 | Memory bandwidth | string / int
3 | TPU enabled | bool
3 | TPU gen | bool
3 | TPU cores | int

**Additional Information**
![GPU drawio](https://github.com/golemfactory/ya-runtime-vm/assets/126684244/4ba56a7a-4a24-4be6-b046-4aa828615c25)

**Comment by evik42** [comment]
added some comments

**Comment by golmek** [comment]
Not totally sure if that will be helpful but maybe some of the above information can be obtained (for Nvidia GPU's) via nvidia-smi (https://developer.nvidia.com/nvidia-system-management-interface) 🤔 ?

**Comment by evik42** [comment]
> Not totally sure if that will be helpful but maybe some of the above information can be obtained (for Nvidia GPU's) via nvidia-smi (https://developer.nvidia.com/nvidia-system-management-interface) 🤔 ?

Many of this information can be accessed in a platform independent way and that is what we should aim for. 
So we could rely on [OpenCL](https://registry.khronos.org/OpenCL/sdk/1.0/docs/man/xhtml/clGetDeviceInfo.html) or [Vulkan API](https://registry.khronos.org/vulkan/specs/1.3-extensions/man/html/VkPhysicalDeviceProperties.html).

**Comment by evik42** [comment]
Depends on #161

**Comment by MrDarthShoe** [comment]
@golmek : Decision for now is to use `nvidia-smi` and we will treat "[OpenCL](https://registry.khronos.org/OpenCL/sdk/1.0/docs/man/xhtml/clGetDeviceInfo.html) or [Vulkan API](https://registry.khronos.org/vulkan/specs/1.3-extensions/man/html/VkPhysicalDeviceProperties.html)."  as a plan B

**Comment by staszek-krotki** [comment]
changes can affect: https://github.com/golemfactory/golem-architecture/pull/86

**Comment by kamirr** [comment]
Up-to-date table of properties to expose:
Information | Type | Unit | Status | Comment
-- | -- | -- | -- | --
GPU model | string | - | ✅ |
CUDA enabled | bool | - | 🚧 | 
FB memory | int | bytes | ✅ | vRAM
CUDA version | int | - | ✅ | This is the version running on the test image -- the only one we can guarantee to work. Speculating about other supported versions (possibly using the GPU model) is left to the user.
CUDA cores | int | -  | ✅ |
Streaming Multiprocessor clock | int | Hz | ✅ |
Graphics clock | int | Hz | ✅ |
Video clock | int | Hz | ✅ |
Memory clock | int | Hz | ✅ |
Memory bus | int | bytes | ✅ |
Memory bandwidth | int | byte / second | ✅ |

**Comment by pwalski** [comment]
The issue with hanging `ya-runtime-vm offer-template` is not related to the drivers, disk/fs.
It can be reproduced using following self-check app doing some `sleep`s:
```
use std::time::Duration;

fn main() -> anyhow::Result<()> {
    println!("{{");
    std::thread::sleep(Duration::from_secs(1));
    println!("\"foo\": \"bar\",");
    std::thread::sleep(Duration::from_secs(1));
    println!("\"moo\": \"meh\"");
    std::thread::sleep(Duration::from_secs(1));
    println!("}}");
    Ok(())
}
```
Here is code handling stdout/err of process in vmrt https://github.com/pwalski/ya-runtime-vm/blob/2dc249c58d713d6fb16d6a2aec64e55b96dfc3d6/runtime/src/self_test.rs#L92
and here https://github.com/pwalski/ya-runtime-vm/blob/2dc249c58d713d6fb16d6a2aec64e55b96dfc3d6/runtime/src/self_test.rs#L162
I will need to debug/fix it.

**Comment by pwalski** [comment]
`ya-runtime-vm test` was failing because `/ya-self-test` stdout was too long.
Handling of `stdout` is broken, but the output was too long because of debug info triggered by the fact number of Cuda cores was 0.
I tried to run `/ya-self-test-gpu` on `ssh` example using image based on `gpu/Dockerfile` on Provider with GPU and I also got 0 Cuda cores.
 Nvidia settings was failing too:
 ```
  nvidia-settings --query all
Unable to init server: Could not connect: Connection refused
```

Is it not necessary to start up Xorg in headless mode before running `nvidia-settings`, and/or configure Xorg.conf and/or setup `DISPLAY` property?
Maybe this task could be unlocked because it can be reproduced both on `ssh` example and on standalone `ya-runtime-vm-nvidia offer-template` on host with GPU?
